### PR TITLE
i#6345: Delay branch instruction in rseq buffer after rollback.

### LINKED ISSUE #6345: Invariant checker reports "Branch target not immediately after branch" error after a CPU ID marker
The invariant checker reports "Branch target not immediately after branch" error when a CPU ID marker is placed in between 
the branch, a timestamp marker, and the branch target.

It should not flag it as an error.

### PR DESCRIPTION
**Pull request by ivankyluk**
When rseq exits on the side, the code rolls back instructions in the rseq buffer. If the last instruction being written from the rseq buffer is a branch, it is not being delayed as in other cases. In the example below, the branch, jbe, is followed by two pairs of timestamp+CPU markers, and then the jmp instruction. The sequence triggers the "Branch target not immediately after branch" invariant error.  

     1508045    ifetch       2 byte(s) @ 0x000000001153345d 76 56                jbe    $0x00000000115334b5 (taken)
     1508046    <marker: timestamp 13340822879775984>
     1508047    <marker: tid 812769 on core 25>
     1508048    <marker: timestamp 13340822879776016>
     1508049    <marker: tid 812769 on core 25>
     1508050    ifetch       5 byte(s) @ 0x00000000115334b5 e9 61 ff ff ff       jmp    $0x000000001153341b <<<<<

The change in this PR delays the branch instruction after rseq rollback, so that it will be written after the timestamp+cpu markers.

Fixes #6345

### CODE REVIEW

File: clients/drcachesim/tests/invariant_checker_test.cpp [hunk 1]
```diff
@@ -176,20 +176,32 @@ bool
...
-            gen_instr(1, 3), gen_marker(2, TRACE_MARKER_TYPE_TIMESTAMP, 0),
-            gen_instr(2, 1),
+            gen_instr(TID, 1),     gen_branch(TID, 2),
+            gen_instr(TID, 3),     gen_marker(TID + 1, TRACE_MARKER_TYPE_TIMESTAMP, 0),
+            gen_instr(TID + 1, 1),
+        };
+        if (!run_checker(memrefs, false))
+            return false;
+    }
+    // Correctly skip CPU ID marker.
```
Comment (derekbruening) [inline clients/drcachesim/tests/invariant_checker_test.cpp:191]: Let's have a case where the branch is taken across the CPU_ID too to ensure the decoding path works there.

File: clients/drcachesim/tests/invariant_checker_test.cpp [hunk 2]
```diff
@@ -176,20 +176,32 @@ bool
 check_branch_target_after_branch()
```
Comment (derekbruening) [inline clients/drcachesim/tests/invariant_checker_test.cpp:176]: > i#6345: Allow CPU ID marker between a branch and it's target

grammar: s/it's/its/

Comment (derekbruening) [inline clients/drcachesim/tests/invariant_checker_test.cpp:176]: > a branch, a timestamp maker, and it's target.

grammar: s/it's/its/

File: clients/drcachesim/tests/invariant_checker_test.cpp [hunk 3]
```diff
@@ -176,20 +176,32 @@ bool
...
+        };
+        if (!run_checker(memrefs, false))
+            return false;
+    }
+    // Correctly skip CPU ID marker.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_instr(TID, 1),
+            gen_branch(TID, 2),
+            gen_marker(TID, TRACE_MARKER_TYPE_TIMESTAMP, TIMESTAMP),
```
Comment (derekbruening) [inline clients/drcachesim/tests/invariant_checker_test.cpp:196]: Wait: the branch is supposed to be delayed across the timestamp+cpuid.  So this shouldn't happen.  We have raw2trace unit tests of this type of thing -- so is there some corner case where you saw this in the wild?

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 4]
```diff
@@ -618,6 +618,10 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
                     // between prev_instr_ and memref.
                     shard->prev_xfer_marker_.marker.marker_type ==
                         TRACE_MARKER_TYPE_KERNEL_EVENT ||
+                    // Allow a CPU ID marker between a branch, and the branch
+                    // target.
+                    (shard->prev_entry_.marker.type == TRACE_TYPE_MARKER &&
+                     shard->prev_entry_.marker.marker_type == TRACE_MARKER_TYPE_CPU_ID) ||
```
Comment (derekbruening) [inline clients/drcachesim/tools/invariant_checker.cpp:624]: The current check is all about 'saw_timestamp_but_no_instr_`.  Relaxing on a cpuid seems to defeat the purpose of the original check as cpuid is pretty much always after a timestamp.  It seems that either this whole check needs to be re-thought and done in a different way, or -- and this seems the case to me -- delaying branches is supposed to never hit and it seems like a bug in raw2trace.

**Review by derekbruening — CHANGES_REQUESTED** [review]
Looking at this more closely: I don't think we shoud relax the check; instead we need to figure out why raw2trace isn't delaying the branch.

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 1 above)
Comment (ivankyluk) [reply clients/drcachesim/tests/invariant_checker_test.cpp]: I've updated the PR to reflect the new approach. PTAL.

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 2 above)
Comment (ivankyluk) [reply clients/drcachesim/tests/invariant_checker_test.cpp]: I've updated the description to reflect the new changes.

Comment (ivankyluk) [reply clients/drcachesim/tests/invariant_checker_test.cpp]: The code has been removed.

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 3 above)
Comment (ivankyluk) [reply clients/drcachesim/tests/invariant_checker_test.cpp]: The corner case is the rollback after rseq side exit. I've updated to the PR accordingly.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 4 above)
Comment (ivankyluk) [reply clients/drcachesim/tools/invariant_checker.cpp]: It's indeed a corner case not being handled in raw2trace. I've updated the PR accordingly.

File: clients/drcachesim/tests/raw2trace_unit_tests.cpp [hunk 5]
```diff
@@ -1508,6 +1508,82 @@ test_rseq_rollback_with_signal(void *drcontext)
         check_entry(entries, idx, TRACE_TYPE_FOOTER, -1));
```
Comment (derekbruening) [inline clients/drcachesim/tests/raw2trace_unit_tests.cpp:1508]: > i#6345: Delay branch instruction in rseq bufffer after rollback

spelling: "buffer"

File: clients/drcachesim/tests/raw2trace_unit_tests.cpp [hunk 6]
```diff
@@ -1508,6 +1508,82 @@ test_rseq_rollback_with_signal(void *drcontext)
...
 }
 
+/* Tests rseq rollback with taken branch right before signal. */
+bool
+test_rseq_rollback_with_signal_after_branch(void *drcontext)
+{
+    std::cerr << "\n===============\nTesting legacy rseq rollback with taken branch "
```
Comment (derekbruening) [inline clients/drcachesim/tests/raw2trace_unit_tests.cpp:1515]: Delete "legacy" as this does have the rseq entry marker.

File: clients/drcachesim/tests/raw2trace_unit_tests.cpp [hunk 7]
```diff
@@ -1508,6 +1508,82 @@ test_rseq_rollback_with_signal(void *drcontext)
...
+    raw.push_back(make_tid());
+    raw.push_back(make_pid());
+    raw.push_back(make_line_size());
+    raw.push_back(make_timestamp());
+    raw.push_back(make_core());
+    raw.push_back(make_marker(TRACE_MARKER_TYPE_RSEQ_ENTRY, offs_move2));
+    raw.push_back(make_block(offs_jcc, 1));
+    raw.push_back(make_marker(TRACE_MARKER_TYPE_RSEQ_ABORT, offs_store));
+    raw.push_back(make_marker(TRACE_MARKER_TYPE_KERNEL_EVENT, offs_store));
+    raw.push_back(make_block(offs_move2, 1));
```
Comment (derekbruening) [inline clients/drcachesim/tests/raw2trace_unit_tests.cpp:1550]: It is confusing to have the rseq abort due to signal followed by the post-rseq instr: normally it would go to a handler with another KERNEL_EVENT?  If the user marked it SIG_IGN: it wouldn't have interrupted at all.  Wouldn't it always go to a handler then?

I see some other tests doing this: at least add a comment saying this is unusual but simplfiied for test purposes?

File: clients/drcachesim/tests/raw2trace_unit_tests.cpp [hunk 8]
```diff
@@ -1508,6 +1508,82 @@ test_rseq_rollback_with_signal(void *drcontext)
...
+        check_entry(entries, idx, TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_TIMESTAMP) &&
+        check_entry(entries, idx, TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CPU_ID) &&
+        check_entry(entries, idx, TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_RSEQ_ENTRY) &&
+        check_entry(entries, idx, TRACE_TYPE_ENCODING, -1) &&
+#ifdef X86_32
+        // An extra encoding entry is needed.
+        check_entry(entries, idx, TRACE_TYPE_ENCODING, -1) &&
+#endif
+        check_entry(entries, idx, TRACE_TYPE_INSTR_TAKEN_JUMP, -1, offs_jcc) &&
+        // The committing store should not be here.
```
Comment (derekbruening) [inline clients/drcachesim/tests/raw2trace_unit_tests.cpp:1577]: Nor the move1

File: clients/drcachesim/tests/raw2trace_unit_tests.cpp [hunk 9]
```diff
@@ -1984,6 +2060,105 @@ test_rseq_side_exit_inverted_with_timestamp(void *drcontext)
...
+    raw.push_back(make_tid());
+    raw.push_back(make_pid());
+    raw.push_back(make_line_size());
+    raw.push_back(make_timestamp(/*value=*/1));
+    raw.push_back(make_core());
+    raw.push_back(make_marker(TRACE_MARKER_TYPE_RSEQ_ENTRY, offs_move2));
+    raw.push_back(make_block(offs_jcc, /*instr_count=*/1));
+    raw.push_back(make_block(offs_move1, /*instr_count=*/2));
+    raw.push_back(make_memref(/*addr=*/42));
+    raw.push_back(make_block(offs_jmp1, /*instr_count=*/1));
```
Comment (derekbruening) [inline clients/drcachesim/tests/raw2trace_unit_tests.cpp:2111]: Add some comments explaining what happened: the jcc actually skipped move1;store;move2?

File: clients/drcachesim/tests/raw2trace_unit_tests.cpp [hunk 10]
```diff
@@ -1984,6 +2060,105 @@ test_rseq_side_exit_inverted_with_timestamp(void *drcontext)
...
+        check_entry(entries, idx, TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_TIMESTAMP) &&
+        check_entry(entries, idx, TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CPU_ID) &&
+        // The jcc instr.
+        check_entry(entries, idx, TRACE_TYPE_ENCODING, -1) &&
+#ifdef X86_32
+        // An extra encoding entry is needed.
+        check_entry(entries, idx, TRACE_TYPE_ENCODING, -1) &&
+#endif
+        check_entry(entries, idx, TRACE_TYPE_INSTR_TAKEN_JUMP, -1, offs_jcc) &&
+        // The move2 + committing store should be gone.
```
Comment (derekbruening) [inline clients/drcachesim/tests/raw2trace_unit_tests.cpp:2148]: move1 too

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 11]
```diff
@@ -841,6 +841,8 @@ raw2trace_t::process_marker_additionally(raw2trace_thread_data_t *tdata,
         log(4, "Rseq abort %d\n", tdata->rseq_past_end_);
         if (!adjust_and_emit_rseq_buffer(tdata, marker_val, marker_val))
             return false;
+        if (!append_delayed_branch(tdata, reinterpret_cast<app_pc>(marker_val)))
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp:844]: Should adjust_and_emit_rseq_buffer call this rather than all its callers having to remember to do so?

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 12]
```diff
@@ -2573,6 +2583,34 @@ raw2trace_t::rollback_rseq_buffer(raw2trace_thread_data_t *tdata,
         if (type_is_instr(static_cast<trace_type_t>(tdata->rseq_buffer_[i].type)))
             ++decode_end;
     }
+    // If the last instructoin after the rollback is a branch, delay the branch.
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp]: spelling: instruction

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 13]
```diff
@@ -2731,10 +2769,16 @@ raw2trace_t::adjust_and_emit_rseq_buffer(raw2trace_thread_data_t *tdata, addr_t
                 // encoding out so it will change back if we see this PC again.
                 rollback_last_encoding(tdata);
             }
-            for (trace_entry_t *e = toadd; e < buf; e++)
-                tdata->rseq_buffer_.push_back(*e);
-            tdata->rseq_buffer_.push_back(jump);
-            tdata->rseq_decode_pcs_.push_back(encoding);
+            if (delayed_branches_exist(tdata)) {
+                write_delayed_branches(tdata, toadd, buf, nullptr, nullptr);
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.cpp:2773]: Is there a test of this, with a synthetic jump added and delayed?

**Review by derekbruening — APPROVED** [review]

File: clients/drcachesim/tests/raw2trace_unit_tests.cpp (hunk 5 above)
Comment (ivankyluk) [reply clients/drcachesim/tests/raw2trace_unit_tests.cpp:1508]: fixed.

File: clients/drcachesim/tests/raw2trace_unit_tests.cpp (hunk 6 above)
Comment (ivankyluk) [reply clients/drcachesim/tests/raw2trace_unit_tests.cpp:1515]: fixed.

File: clients/drcachesim/tests/raw2trace_unit_tests.cpp (hunk 7 above)
Comment (ivankyluk) [reply clients/drcachesim/tests/raw2trace_unit_tests.cpp:1550]: Yeah, it's a simplified version of traces from code_api|tool.drcacheoff.rseq:

      141487      105678:     3662080 ifetch       7 byte(s) @ 0x00007fe7abdfa938 80 3d b9 38 00 00 00 cmp    <rel> 0x00007fe7abdfe1f8, $0x00
      141488      105678:     3662080 read         1 byte(s) @ 0x00007fe7abdfe1f8 by PC 0x00007fe7abdfa938
      141489      105679:     3662080 ifetch       2 byte(s) @ 0x00007fe7abdfa93f 74 0a                jz     $0x00007fe7abdfa94b (untaken)
      141490      105680:     3662080 ifetch       8 byte(s) @ 0x00007fe7abdfa941 0f 18 1c 25 01 00 00 prefetcht2 0x01
      141490      105680:     3662080                                             00
      141491      105680:     3662080 pref-r-L3    1 byte(s) @ 0x0000000000000001 by PC 0x00007fe7abdfa941
      141492      105681:     3662080 ifetch       2 byte(s) @ 0x00007fe7abdfa949 0f 0b                ud2
      141493      105681:     3662080 <marker: rseq abort from 0x7fe7abdfa949 to handler>
      141494      105681:     3662080 <marker: kernel xfer from 0x7fe7abdfa949 to handler>
      141495      105681:     3662080 <marker: timestamp 13351125423049607>
      141496      105681:     3662080 <marker: tid 3662080 on core 4110>
      141497      105681:     3662080 <marker: kernel xfer from 0x7fe7abdfa957 to handler>
      141498      105681:     3662080 <marker: timestamp 13351125423049609>
      141499      105681:     3662080 <marker: tid 3662080 on core 4110>

I've added a comment accordingly.

File: clients/drcachesim/tests/raw2trace_unit_tests.cpp (hunk 8 above)
Comment (ivankyluk) [reply clients/drcachesim/tests/raw2trace_unit_tests.cpp:1577]: done.

File: clients/drcachesim/tests/raw2trace_unit_tests.cpp (hunk 9 above)
Comment (ivankyluk) [reply clients/drcachesim/tests/raw2trace_unit_tests.cpp:2111]: done.

File: clients/drcachesim/tests/raw2trace_unit_tests.cpp (hunk 10 above)
Comment (ivankyluk) [reply clients/drcachesim/tests/raw2trace_unit_tests.cpp:2148]: done.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 12 above)
Comment (ivankyluk) [reply clients/drcachesim/tracer/raw2trace.cpp]: done.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 11 above)
Comment (ivankyluk) [reply clients/drcachesim/tracer/raw2trace.cpp:844]: In append_bb_entries, the call to append_delayed_branch depends on whether the instruction is a branch or not, i.e. line 1961

           if (tdata->rseq_past_end_) {
                if (!adjust_and_emit_rseq_buffer(tdata, instr_pc))
                    return false;
                if (!instr->is_cti()) {
                    // Write out delayed branches now that we have a target.
                    if (!append_delayed_branch(tdata, orig_pc))
                        return false;
                }

which is difficult to generalize and add to adjust_and_emit_rseq_buffer.

File: clients/drcachesim/tracer/raw2trace.cpp (hunk 13 above)
Comment (ivankyluk) [reply clients/drcachesim/tracer/raw2trace.cpp:2773]: I've added a new test to cover it.
